File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/practice/Subsets/solution.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        # 메서드
        def backTracking(result, temp, start):
            result.append(temp[:])

            for i in range(start, len(nums)):
                temp.append(nums[i])
                backTracking(result, temp, i + 1)
                temp.pop()

        # 메인 로직
        result = []
        backTracking(result, [], 0)

        return result

File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/practice/Subsets/Solution.java
import java.util.*;

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<Integer> subset = new ArrayList<>();
        List<List<Integer>> answer = new ArrayList<>();
        answer.add(new ArrayList<>());
        backTracking(nums, 0, subset, answer);

        return answer;
    }

    private void backTracking(int[] nums, int start, List<Integer> subset, List<List<Integer>> answer) {
        for (int i = start; i < nums.length; i++) {
            subset.add(nums[i]);
            answer.add(new ArrayList<>(subset));
            backTracking(nums, i + 1, subset, answer);
            subset.remove(subset.size() - 1);
        }
    }
}
